### DISCUSSION TITLE
(De)serialization of deeper field

### DISCUSSION THREAD

**Discussion by Rogdham** [wyfo/apischema#372] [Q&A]
Hello,

I'm trying `apischema` out, and was wondering if there is an easy way to serialize/deserialize nested data, something like this:
```py
data = {
    "filename": "abc.py",
    "lines": {
        "start": 13,
        "end": 37,
    },
}

@dataclass
class CodeSnippet:
    filename: str
    start_line: int  # add some field metadata here I guess
    end_line: int    # same here

obj = CodeSnippet("abc.py", 13, 37)
assert deserialize(CodeSnippet, data) == obj
assert serialize(CodeSnippet, obj) == data
```

I made some progress and got the following code working for serialization:
```py

@dataclass
class CodeSnippet:
    filename: str
    start_line: int = field(metadata=skip)
    end_line: int = field(metadata=skip)

    @serialized
    @property
    def lines(self) -> dict:
        return {
            "start": self.start_line,
            "end": self.end_line,
        }
```

I guess I could go this route and add deserialization in a similar manner, but I'm new to `apischema` and wanted to know if there was an easier way, e.g. something like `alias(["lines", "start"])`.

**Comment by wyfo** [comment]
Hi,
I think the current *apischema* way of handling your case is to use [conversions](https://wyfo.github.io/apischema/0.17/conversions/):
```python
from dataclasses import dataclass

from apischema import deserialize, deserializer, serialize, serializer


@dataclass
class CodeSnippet:
    filename: str
    start_line: int 
    end_line: int


@dataclass
class NestedCodeSnippet:
    filename: str

    @dataclass
    class Lines:
        start: int
        end: int

    lines: Lines


@deserializer
def from_nested(nested: NestedCodeSnippet) -> CodeSnippet:
    return CodeSnippet(nested.filename, nested.lines.start, nested.lines.end)


@serializer
def to_nested(cs: CodeSnippet) -> NestedCodeSnippet:
    return NestedCodeSnippet(
        cs.filename, NestedCodeSnippet.Lines(cs.start_line, cs.end_line)
    )


data = {"filename": "abc.py", "lines": {"start": 13, "end": 37}}
obj = CodeSnippet("abc.py", 13, 37)
assert deserialize(CodeSnippet, data) == obj
assert serialize(CodeSnippet, obj) == data
```

But I know that in the case where there is a lot of fields in `CodeSnippet`, redeclaring all of them in `NestedCodeSnippet` can be tedious and frustrating. Sharing common fields in a parent class could be a solution to this problem, but I'm not so fan of having changing code structure because of serialization; apischema aims to be purely additive to existing code.

Your idea with `alias` is interesting, and implementing it seems even more interesting. I will surely work on it this week.
